Issue 3043: Switch to Spotbugs

eolivelli opened the pull request:
Switch to Spotbugs, the successor of FindBugs
    - replace Findbugs with Spotbugs
    - fix minor issues reported by Spotbugs
   -  fix a bug in CommitRequestHandler
   

Signed-off-by: Enrico Olivelli <eolivelli@apache.org>

**Change log description**  
Remove Findbugs plugin and add Spotbugs plugin

**Purpose of the change**  
Fixes #3043

**What the code does**  
- Change Gradle configuration
- fix issues reported by Spotbugs

**How to verify it**  
Run automated checks like
./gradlew findbugsMain findbugsTest

eolivelli commented:
@fpj @tkaitchuck @shrids 
This change is needed to JDK10/11 compatibility, Findbugs does not work

eolivelli commented on gradle/spotbugs.gradle:
```diff
@@ -0,0 +1,21 @@
+        effort = "default"        
+        includeFilter = file("$rootDir/checkstyle/spotbugs-include.xml")
+        excludeFilter = file("$rootDir/checkstyle/spotbugs-exclude.xml")
+
```
This changed a bit, but on my laptop Spotbugs is working.
I don't know how much was important the HTML report.

tkaitchuck replied:
The HTML report is important for human readability. So people can locate where the problem is.

eolivelli commented on gradle/java.gradle:
```diff
@@ -10,12 +10,15 @@
     apply plugin: 'maven'
 
     compileJava {
+        sourceCompatibility = "8"
```
this is not strictly needed, I can move this in another commit, but it better to have Java version explicitly

eolivelli commented:
Thanks for your quick feedback

There is still a reference to findbugs annotations.
I will remove it in favour of the same jar but with spotbugs groupid/artifact id.

The include/exclude files are using the same format of Findbugs so they use the FindBugs XML tags. This is expected

@andreipaduroiu I have removed the html report, was it important? Now the report is in XML format. Is it okay?

andreipaduroiu commented:
I don't particularly care if it's HTML or XML as long as it tells me where the problem is and how I should fix it. Findbugs was doing that (for the most part; sometimes I still had to get creative in addressing whatever problem it found).

eolivelli commented:
Pushed the migration of the findbugs annotations package according to:
https://spotbugs.readthedocs.io/en/stable/migration.html

eolivelli commented:
@andreipaduroiu After switching the annotations Spotbugs signalled another issue on CommitRequestHandler which I think it could be a real bug.

eolivelli commented:
@andreipaduroiu I have addressed your comments. The word "findbugs" cannot be dropped because Spotbugs is a fork of FindBugs and so it shares the same package names and XML file dialects.
Lombok treats Spotbugs as Findbugs.

I have also tried to move Spotbugs annotations to 'compileOnly' scope but the compilation will fail, I think this is due to Lombok. Anyway Findbugs was already in 'compile' scope.

The patch is ready

tkaitchuck commented on checkstyle/spotbugs-exclude.xml:
```diff
@@ -44,6 +47,7 @@
             <Bug category="PERFORMANCE" />
             <Bug pattern="URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD" />
             <Bug pattern="DMI_RANDOM_USED_ONLY_ONCE" />
+            <Bug pattern="FE_FLOATING_POINT_EQUALITY" />
```
I'm not sure we want this one. It sounds like a way to get flaky tests. Perhaps a utility function in Asserts would help?

eolivelli replied:
Okay, I will search where spotbugs was pointing this issue and replace with a new function in AssertExtensions.
Junit already has special floating point equality assertions, I can't remember why I decided to suppress the warning and not to use junit builtin support. I will be back with a better answer

eolivelli replied:
@tkaitchuck  done, I have added a new method to AssertExtensions and removed the exclusion

tkaitchuck commented on controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java:
```diff
@@ -145,7 +145,7 @@ public CommitRequestHandler(final StreamMetadataStore streamMetadataStore,
                                     // if state is sealing, we should continue with commit so that we allow for completion of transactions
                                     // in commit state.
-                                    if (state.equals(State.SEALING)) {
+                                    if (state.getObject().equals(State.SEALING)) {
```
This looks like it would change the behavior. Is this a bug that was caught?

eolivelli replied:
Yes, this is a bug found by spotbugs, and apparently this condition is not covered by any test.
I am still new to this code, so I did not add a test case, but the bug is clear to me just by looking at the code

shiveshr replied:
this is a good catch. 
added issue to add unit test for this scenario https://github.com/pravega/pravega/issues/3122

eolivelli commented:
@tkaitchuck I am still new to gradle, I did not find any working way to configure the plugin in order to produce an HTML, docs are really skinny, and they just say that the config is the same as the findbugs plugin but it is not true.
I will try again, fortunately grade build scripts are in pure groovy, so some way it is possible to debug.
If anyone in the group can give me an hint I will appreciate it very much

eolivelli commented:
@tkaitchuck actually spotbugs gradle plugin cannot produce HTML reports,
see https://github.com/spotbugs/spotbugs-gradle-plugin/issues/32

the XML report is anyway useful

eolivelli commented:
@tkaitchuck I have addresses all of your comments and answers. PTAL

eolivelli commented:
@fpj Checkstyle should be okay now
